router/peers_test.go

```diff
@@ -0,0 +1,170 @@
+	}
+	connsIterator(connsBuf, func(remoteNameByte []byte, _ string) {
+		remoteName := PeerNameFromBin(remoteNameByte)
+		if _, found := checkConns[remoteName]; found {
+			delete(checkConns, remoteName)
+		} else {
+			t.Fatalf("%s: Unexpected connection decoded from %s to %s", wt.CallSite(5), ourName, remoteName)
+		}
+	})
+	if len(checkConns) > 0 {
+		t.Fatalf("%s: Expected connections not found: from %s to %v", wt.CallSite(3), ourName, checkConns)
+	}
+}
+
+func decodePeerInfo(t *testing.T, decoder *gob.Decoder) []*decodedPeerInfo {
+	peerInfo := make([]*decodedPeerInfo, 0)
+	for {
+		nameByte, uid, version, connsBuf, decErr := decodePeerNoConns(decoder)
+		if decErr == io.EOF {
+			break
+		} else if decErr != nil {
+			t.Fatalf("%s: Error when decoding peer (%s)", wt.CallSite(2), decErr)
+		}
+		peerInfo = append(peerInfo, &decodedPeerInfo{PeerNameFromBin(nameByte), uid, version, connsBuf})
+	}
+	return peerInfo
+}
+
+func checkBlank(t *testing.T, update []byte) {
+	decoder := gob.NewDecoder(bytes.NewReader(update))
+	peerInfo := decodePeerInfo(t, decoder)
+	if len(peerInfo) != 0 {
+		t.Fatalf("%s: Expected 0 items but got %s", wt.CallSite(2), peerInfo)
+	}
+}
+
+func checkEncoding(t *testing.T, update []byte, routers []*Router, connections [][]Connection) {
+	decoder := gob.NewDecoder(bytes.NewReader(update))
+
+	// Peers can come either way round, so read them all in and sort them
+	peerInfo := decodePeerInfo(t, decoder)
+	sort.Sort(byName(peerInfo))
+	N := len(peerInfo)
+	if N != len(routers) {
+		t.Fatalf("%s: Expected %d items but got %d: %s", wt.CallSite(2), len(routers), N, peerInfo)
+	}
+	for i := 0; i < N; i++ {
+		checkPeerDetails(t, peerInfo[i], routers[i].Ourself.Peer)
+		checkConnsEncoding(t, peerInfo[i].name, peerInfo[i].connsBuf, connections[i])
+	}
+}
+
+func TestPeersEncoding(t *testing.T) {
+	const (
+		peer1NameString = "01:00:00:01:00:00"
+		peer2NameString = "02:00:00:02:00:00"
+		peer3NameString = "03:00:00:03:00:00"
+	)
+	var (
+		peer1Name, _ = PeerNameFromString(peer1NameString)
+		peer2Name, _ = PeerNameFromString(peer2NameString)
+		peer3Name, _ = PeerNameFromString(peer3NameString)
+	)
+
+	// Create some peers
+	r1 := NewTestRouter(t, peer1Name, nil)
+	r2 := NewTestRouter(t, peer2Name, nil)
+	r3 := NewTestRouter(t, peer3Name, nil)
+
+	// Check state when they have no connections
+	checkEncoding(t, r1.Peers.EncodeAllPeers(), rs(r1), ca(nil))
+	checkEncoding(t, r2.Peers.EncodeAllPeers(), rs(r2), ca(nil))
+
+	// Now try adding some connections
+	r1.AddTestConnection(r2)
+	r2.AddTestConnection(r1)
+	checkEncoding(t, r1.Peers.EncodeAllPeers(), rs(r1, r2), ca(cs(r2), nil))
+	checkEncoding(t, r2.Peers.EncodeAllPeers(), rs(r1, r2), ca(nil, cs(r1)))
+	// Currently, the connection from 2 to 3 is one-way only
+	r2.AddTestConnection(r3)
+	checkEncoding(t, r2.Peers.EncodeAllPeers(), rs(r1, r2, r3), ca(nil, cs(r1, r3), nil))
+	checkEncoding(t, r3.Peers.EncodeAllPeers(), rs(r3), ca(nil))
+}
```

Comment:
That would check that the code is self-consistent, but would not check that the encoding matches the documentation or the previous release.  Since what you suggest is effectively done inside `TestGossipTopology(),` this test can either stand as an additional warning if the protocol is accidentally broken, or deleted if maintaining it is too much of a burden.